eeps/eep-0065.md

```diff
@@ -0,0 +1,282 @@
+
+All of these simple prefixes has to become separate tokens in the scanner
+(lexer): «`bf"`» would constitute the start token for a binary string
+with interpolation syntax.  So would «`bf"""`», «`b"`», «`b"""`»,
+and so on.
+
+The scanner would have to know of all combinations of prefix characters
+and emit distinct tokens for every combination.
+
+Today, the character sequence «`b`», «`f`», «`"`» is scanned as a token
+for the atom «`bf`» followed by the string start token «`"`».
+That combination fails in the parser so it is syntactically invalid today,
+which is what makes simple prefixes a possible language extension.
+
+A simple prefix approach would have to scan a number of characters
+ahead to distinguish between an atom followed by string start
+vs. prefixed string start, and it would be a different number
+of characters depending on which atom characters that have
+been found so far.  This is rather messy.
+
+Furthermore, it is likely that we want the feature of choosing
+[String Delimiters][], especially for regular expressions such as:
+
+    re(^"+.*/.*$)
+
+Among the desired delimiters are «`/`» and «`<`»+«`>`».  The currently
+valid code «`b<X`» meaning atom «`b`» less than «`X`», would instead
+have to be interpreted as prefixed string start «`b<`» with «`X`»
+being the first string content character.
+
+For the «`/`» character we run into similar problems with for example
+«`b/X`», which would be a run-time error today, but if we also would
+want capital letter prefixes, then «`B/X`» is perfectly valid today
+but would become a string start.
+
+There are more likely problems with simple string prefixes:
+«`#bf{`» is today the start of a record named «`bf`», and is
+scanned as punctuation character «`#`», atom «`bf`» and separator «`{`»,
+which the parser sorts out to be a record start.
+
+With simple prefix characters the scanner would have to be rewritten
+to recognize «`#bf`» as a new record token, a rewrite that might cause
+unexpected changes in record handling.  For example, today, «`# bf {`»
+is also a valid record start, so to be completely compatible the scanner
+would have to allow white-space or even newlines within the new record
+token, between «`#`» and the atom characters, which would be really ugly...
+
+For reasons, maybe some of them above, Elixir has chosen to use
+the «`~`» character as the start of a string prefix which they
+call a ["Sigil"][1].
+
+### Sigil
+
+In a general sense, a [Sigil][3], is a prefix to a variable
+that indicates its type, such as «`$I`» in Basic or Perl,
+ where «`$`» is the sigil and «`I`» is the variable.
+
+Here we define a Sigil as a prefix to a string literal
+that indicates how to interpret the string literal.
+It consists of:
+
+1. The character «`~`»
+2. An (optional) character in the range «`a`»-«`z`» or «`A`»-«`Z`»,
+   that specifies the [Sigil Type][].  The Sigil Type can
+   affect how the String Content is scanned, for example
+   if escape sequences are honoured.
```

Comment:
Multi-letter Sigils in Elixir, indeed... Good to know!

In Erlang there should be no problem to allow multi-letter sigils without uppercase restriction.

Word lists, time formats and custom sigils are among the ones I think are the least interesting for Erlang.  The ones in the EEP I think are the interesting, and for them (binary, interpolation, and regexp) it would be attractive that the sigil is short:  `~utf8_binary"Content"` is not obviously better than `<<"Content"/utf8>>`, but `~s"Content"` is.

So I don't think those special Sigils are strong arguments to allow multi-letter Sigils.

That said; my first internal suggestion was that a Sigil String Literal would be `~xyz""` with no suffix, and even regexp flags should be handled in the prefix.  And I still see that it would be nice to have interpolation as a character flag in the prefix, making it easier to introduce Sigils before deciding on interpolation.

Still,  see no really good reason to limit the sigil prefix to one character - I only copied what I thought was Elixir's syntax.  But I find more reason to allow multi-letter sigils in order to to have more flag characters in the prefix, than to have a more descriptive name as the sigil prefix.

Reply:
> it would be attractive that the sigil is short: ~utf8_binary"Content" is not obviously better than <<"Content"/utf8>>, but ~s"Content" is.

I agree.

---

It also all depends on what is the extensibility story. If the sigils are all hardcoded in the tokenizer/parser, then you can keep it single-letter and make it to multi-letter when needed. But if you want to allow custom sigils via parse transforms, then allowing the community to be clearer with longer names is a huge plus. And I would say some of the best use cases of sigils in Elixir comes from the community:

1. Phoenix uses sigils for writing routes, such as: `~p"/posts/#{id}"` - It works the same as `~s"/posts/#{id}"` except that Phoenix also validates that the route exists, helping developers catch typos (such as if someone wrote ~p"/post/#{id}" instead).

2. Nx uses `~MAT` and `~VEC` for writing matrices and vectors. Instead of:

```elixir
tensor([
  [1.0, 0.0, 2.0],
  [3.0, 5.0, 4.0],
  [5.0, 4.0, 6.0]
])
```

you can write:

```elixir
~MAT"
  1.0 0.0 2.0
  3.0 5.0 4.0
  5.0 4.0 6.0
"
```

Besides helping with readability, it also converts the tensor into its underlying binary format at compile-time.
You can also specify the type+bit at the end: `~MAT[...]f16`

3. Phoenix uses `~H` (but it may have been `~HTML` if we could go back in time) for writing HTML templates. It supports interpolation and validates the contents is actually valid HTML. You can also see someone using this to write valid XML fragments, etc